### PR TITLE
Api improvements + #10

### PR DESCRIPTION
**Pull request by cambierr**

### CODE REVIEW

File: src/main/java/org/thethingsnetwork/java/app/lib/Message.java [hunk 1]
```diff
@@ -1,39 +1,45 @@
...
- * @author Romain Cambier <romain@shareif.com>
+ * This is a wrapper class for JSONObject to provide support for base64 encoded payload
+ *
+ * @author Romain Cambier
  */
 public class Message extends JSONObject {
 
-    public byte[] getBinary(String _key) {
-        Object object = get(_key);
+    public byte[] getRawPayload() {
```
Comment (FokkeZB) [inline src/main/java/org/thethingsnetwork/java/app/lib/Message.java]: I finally see how JsonObject works very different from other implementations for other languages I've seen and that you have to use a type specific getter and also go [all weirdo](http://stackoverflow.com/a/14751649) to get deep fields. I get why you did `getBinary()` now and that it is probably more Java-like then what you changed it to now. So let's change it back. Sorry 🙃

File: src/main/java/org/thethingsnetwork/java/app/lib/Client.java [hunk 2]
```diff
@@ -70,14 +72,34 @@
...
+
+        URI tempBroker = new URI(_source.contains(".") ? _source : (_source + ".thethings.network"));
+
+        switch (tempBroker.getScheme()) {
+            case "tcp":
+                if (tempBroker.getPort() == -1) {
+                    return tempBroker.toString() + ":1883";
+                }
+                break;
+            case "ssl":
```
Comment (FokkeZB) [inline src/main/java/org/thethingsnetwork/java/app/lib/Client.java]: If it's SSL then probably you need certificates as well. How would you set that? Here's how I've set it up for Nodejs: https://github.com/TheThingsNetwork/node-app-lib/pull/20

File: src/main/java/org/thethingsnetwork/java/app/lib/Client.java [hunk 3]
```diff
@@ -299,5 +321,20 @@ public void send(String _devId, byte[] _payload, int _port) throws MqttException
         data.put("port", _port != 0 ? _port : 1);
         mqttClient.publish(appId + "/devices/" + _devId + "/down", data.toString().getBytes(), 0, false);
     }
+    
+    /**
+     * Send a downlink message using pre-registered encoder
```
Comment (FokkeZB) [inline src/main/java/org/thethingsnetwork/java/app/lib/Client.java]: Great! The Node.js lib now also supports arrays, but we can do that in a later PR if you like: https://github.com/TheThingsNetwork/node-app-lib/pull/17

File: src/main/java/Test.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
+/*
```
Comment (FokkeZB) [inline src/main/java/Test.java]: Could you include steps (in the readme) on:

1. How to build and publish this project (probably link some guide)
2. How to run this test app

File: src/main/java/org/thethingsnetwork/java/app/lib/Message.java (hunk 1 above)
Comment (cambierr) [inline src/main/java/org/thethingsnetwork/java/app/lib/Message.java]: Ok for this, no worry

File: src/main/java/org/thethingsnetwork/java/app/lib/Client.java (hunk 3 above)
Comment (cambierr) [inline src/main/java/org/thethingsnetwork/java/app/lib/Client.java]: Not sure to understand what you mean ? Is anything missing in the java impl ?

File: src/main/java/Test.java (hunk 4 above)
Comment (cambierr) [inline src/main/java/Test.java]: 1. you mean to publish to maven ? Does it really make sense to have that in the readme ?
2. ok

File: src/main/java/org/thethingsnetwork/java/app/lib/Client.java (hunk 2 above)
Comment (cambierr) [inline src/main/java/org/thethingsnetwork/java/app/lib/Client.java]: ok

**Comment by cambierr** [comment]
So, I think all required changes have been done! :-)

**Comment by FokkeZB** [comment]
Thanks @cambierr! I tried to run the test per your instructions by I get a null pointer:

```
[INFO] --- exec-maven-plugin:1.5.0:java (default-cli) @ sample ---
[WARNING]
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.mojo.exec.ExecJavaMojo$1.run(ExecJavaMojo.java:294)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.NullPointerException
	at org.thethingsnetwork.java.app.lib.Client.validateBroker(Client.java:88)
	at org.thethingsnetwork.java.app.lib.Client.<init>(Client.java:78)
	at org.thethingsnetwork.java.app.sample.Test.main(Test.java:42)
	... 6 more
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 1.823 s
[INFO] Finished at: 2016-09-22T08:06:36+02:00
[INFO] Final Memory: 18M/304M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.5.0:java (default-cli) on project sample: An exception occured while executing the Java class. null: InvocationTargetException: NullPointerException -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.5.0:java (default-cli) on project sample: An exception occured while executing the Java class. null
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:212)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:199)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoExecutionException: An exception occured while executing the Java class. null
	at org.codehaus.mojo.exec.ExecJavaMojo.execute(ExecJavaMojo.java:345)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
	... 20 more
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.mojo.exec.ExecJavaMojo$1.run(ExecJavaMojo.java:294)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.NullPointerException
	at org.thethingsnetwork.java.app.lib.Client.validateBroker(Client.java:88)
	at org.thethingsnetwork.java.app.lib.Client.<init>(Client.java:78)
	at org.thethingsnetwork.java.app.sample.Test.main(Test.java:42)
	... 6 more
[ERROR]
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

**Comment by cambierr** [comment]
You took the code from master, right? my instructions will actually be valid once the PR will me merged.
Meanwhile, you should use the code from my branch

**Comment by FokkeZB** [comment]
I did use the latest `api_improvements` branch

**Comment by cambierr** [comment]
@FokkeZB ready for merge ?

**Pull request merged**